logger/handler/ringbuffer.go

```diff
@@ -0,0 +1,149 @@
+package handler
+
+import (
+	"container/ring"
+	"fmt"
+	"log"
+	"net/http"
+	"regexp"
+	"strconv"
+	"sync"
+
+	"github.com/deis/deis/logger/drain"
+	"github.com/deis/deis/logger/syslog"
+)
+
+var logStorage = make(map[string]*ring.Ring)
+var ringBufferSize int
+
+var regexpForPost = regexp.MustCompile(`^/([-a-z0-9]+)/.*`)
+var regexpForGet = regexp.MustCompile(`^/([-a-z0-9]+)/([0-9]+)/.*`)
+
+var mu sync.Mutex
+
+// Add log message to main map with ring byffers by project name
+func addToStorage(name string, message string) {
+	mu.Lock()
+	defer mu.Unlock()
+	currentRing, ok := logStorage[name]
+	if !ok {
+		r := ring.New(ringBufferSize)
+		r.Value = message
+		logStorage[name] = r
+	} else {
+		r := currentRing.Next()
+		r.Value = message
+		logStorage[name] = r
+	}
+}
+
+// Get specific amount of log lines for application name from main map with ring buffers
+func getFromStorage(name string, lines int) string {
+	mu.Lock()
+	defer mu.Unlock()
+	currentRing, ok := logStorage[name]
+	if !ok {
+		return fmt.Sprintf("Could not find logs for project '%s'", name)
+	}
+	var data string
+	getLine := func(line interface{}) {
+		if line == nil || lines <= 0 {
+			return
+		}
+		lines -= 1
+		if lines == 0 {
+			data += fmt.Sprint(line)
+		} else {
+			data += fmt.Sprintln(line)
+		}
+	}
+
+	currentRing.Next().Do(getLine)
+	return data
+}
+
+func RingBufferHandler(bufferSize int, webServicePort int) *Handler {
+	ringBufferSize = bufferSize
+	h := Handler{
+		BaseHandler: syslog.NewBaseHandler(5, filter, false),
+	}
+	go h.ringBufferLoop()
+	go startWebService(webServicePort)
+	fmt.Printf("Web service started on %d port\n", webServicePort)
+	return &h
+}
+
+// Main loop for this handler, each log line will be sended to drain (if DrainURI specified) and copied to log storage
+func (h *Handler) ringBufferLoop() {
+	var err error
+	var projectName string
+	var message string
```

Comment:
you don't need any of these declared as they are overwritten on each loop. remove them and add the `:=` to where they are being assigned to below.